lib/core/src/Cardano/Wallet.hs

```diff
@@ -692,34 +701,41 @@ newWalletLayer tracer bp db nw tl = do
                         then Ready
                         else Restoring progress'
 
-            let txs_cps = applyBlocks @s @t (NE.toList blocks) cp
-            let txs = fold $ fst <$> txs_cps
-
-            let cpLast = fromMaybe cp $ lastMay $ snd <$> txs_cps
-            let (Quantity bhLast) = blockHeight cpLast
+            -- NOTE:
+            -- Always store the last checkpoint from the batch and all new
+            -- transactions.
+            let cpLast = NE.last cps
+            let Quantity bhLast = blockHeight cpLast
+            let meta' = calculateMetadata (view #slotId $ currentTip cpLast)
+            DB.putCheckpoint db (PrimaryKey wid) cpLast
+            DB.putTxHistory db (PrimaryKey wid) newTxs
+            DB.putWalletMeta db (PrimaryKey wid) meta'
+
+            -- NOTE (1):
+            -- We cast `k` and `nodeHeight` to 'Integer' since at a given point
+            -- in time, `k` may be greater than the tip.
+            let (Quantity k) = epochStability
+            let bhUnstable :: Integer
+                bhUnstable = fromIntegral nodeHeight - fromIntegral k
+            forM_ (NE.init cps) $ \cp -> do
```

Comment:
Would it be possible to define a separate pure function that filters `cps` into a batch? Then this can more easily be tested. Something like:
```haskell
let cps' = batchCheckpoints k networkHeight cps
mapM (DB.putCheckpoint db (PrimaryKey wid)) cps'
```

Reply:
Makes sense. Maybe in the form of an iterator "forEachUnstable action"..